pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java

```diff
@@ -0,0 +1,356 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.migrating;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.jaxen.JaxenException;
+
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.java.ast.ASTExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTForInit;
+import net.sourceforge.pmd.lang.java.ast.ASTForStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTForUpdate;
+import net.sourceforge.pmd.lang.java.ast.ASTName;
+import net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix;
+import net.sourceforge.pmd.lang.java.ast.ASTPrimarySuffix;
+import net.sourceforge.pmd.lang.java.ast.ASTRelationalExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclarator;
+import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclaratorId;
+import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
+import net.sourceforge.pmd.lang.java.symboltable.VariableNameDeclaration;
+import net.sourceforge.pmd.lang.symboltable.NameOccurrence;
+import net.sourceforge.pmd.lang.symboltable.Scope;
+
+/**
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+public class ForLoopCanBeForeachRule extends AbstractJavaRule {
+
+    @Override
+    public Object visit(ASTForStatement node, Object data) {
+
+        final ASTForInit init = node.getFirstChildOfType(ASTForInit.class);
+        final ASTForUpdate update = node.getFirstChildOfType(ASTForUpdate.class);
+        final ASTExpression guardCondition = node.getFirstChildOfType(ASTExpression.class);
+
+        if (init == null && update == null || guardCondition == null) {
+            return super.visit(node, data);
+        }
+
+        Entry<VariableNameDeclaration, List<NameOccurrence>> indexDecl = getIndexVarDeclaration(init, update);
+
+        if (indexDecl == null) {
+            return super.visit(node, data);
+        }
+
+
+        List<NameOccurrence> occurrences = indexDecl.getValue();
+        VariableNameDeclaration index = indexDecl.getKey();
+
+        if ("Iterator".equals(index.getTypeImage()) && isReplaceableIteratorLoop(indexDecl, guardCondition, update)) {
```

Comment:
There is a `net.sourceforge.pmd.lang.java.typeresolution.TypeHelper` utility class, you could use here:

    TypeHelper.isA(index, Iterator.class)

It uses typeresolution, if the type is available and as a fallback the type image.
The we should of course enable typeresolution for this rule.